src/libraries/System.Security.Cryptography.Encoding/tests/Cbor/CborReader.String.cs

```diff
@@ -120,5 +148,145 @@ public void ReadEndByteString()
+            ReadOnlySpan<byte> buffer = _buffer.Span;
+
+            int concatenatedStringSize = 0;
+            foreach ((int o, int l) in ranges)
+            {
+                concatenatedStringSize += s_utf8Encoding.GetCharCount(buffer.Slice(o, l));
+            }
+
+            if (concatenatedStringSize > destination.Length)
+            {
+                charsWritten = 0;
+                return false;
+            }
+
+            foreach ((int o, int l) in ranges)
+            {
+                s_utf8Encoding.GetChars(buffer.Slice(o, l), destination);
+                destination = destination.Slice(l);
+            }
+
+            charsWritten = concatenatedStringSize;
+            AdvanceBuffer(offset);
+            DecrementRemainingItemCount();
+            return true;
+        }
+
+        private byte[] ReadChunkedByteStringConcatenated()
+        {
+            List<(int offset, int length)> ranges = ReadChunkedStringRanges(CborMajorType.ByteString, out int offset, out int concatenatedBufferSize);
+            var output = new byte[concatenatedBufferSize];
+
+            ReadOnlySpan<byte> source = _buffer.Span;
+            Span<byte> target = output;
+
+            foreach ((int o, int l) in ranges)
+            {
+                source.Slice(o, l).CopyTo(target);
+                target = target.Slice(l);
+            }
+
+            Debug.Assert(target.IsEmpty);
+            AdvanceBuffer(offset);
+            DecrementRemainingItemCount();
+            return output;
+        }
+
+        private string ReadChunkedTextStringConcatenated()
+        {
+            List<(int offset, int length)> ranges = ReadChunkedStringRanges(CborMajorType.TextString, out int offset, out int concatenatedBufferSize);
+            ReadOnlySpan<byte> buffer = _buffer.Span;
+            int concatenatedStringSize = 0;
+
+            foreach ((int o, int l) in ranges)
+            {
+                concatenatedStringSize += s_utf8Encoding.GetCharCount(buffer.Slice(o, l));
+            }
+
+            var output = new char[concatenatedStringSize];
+            Span<char> target = output;
+
+            foreach ((int o, int l) in ranges)
+            {
+                s_utf8Encoding.GetChars(buffer.Slice(o, l), target);
+                target = target.Slice(l);
+            }
+
+            Debug.Assert(target.IsEmpty);
+            AdvanceBuffer(offset);
+            DecrementRemainingItemCount();
+            return new string(output);
+        }
+
+        // reads a buffer starting with an indefinite-length string,
+        // performing validation and returning a list of ranges containing the individual chunk payloads
+        private List<(int offset, int length)> ReadChunkedStringRanges(CborMajorType type, out int encodingLength, out int concatenatedBufferSize)
+        {
+            var ranges = new List<(int offset, int length)>();
```

Comment:
I don't know if it's worth the complexity; but you could do something like

```C#
List<(int offset, int length)> ranges = Interlocked.Exchange(ref _ranges, null);
ranges ??= new List<(int, int)>();
ranges.Clear();
```

And have all of the calling members assign it back to the field right before returning.

It's not super important, though.

Reply:
I'll give it a shot, provided it doesn't make things too hard to read. Or maybe wait until benchmarking demonstrates issues with allocations.

Also, why Interlocked? The class offers no thread safety guarantees.

Reply:
It's both a quick way of "I've taken this" (in case somehow this ever went recursive, which I know isn't currently possible) and it means that the kinds of scary race conditions are limited. (overlapping writes from a read method is scarier to me than read twice or read slightly forward)